// docz-utils/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@babel/parser');
require('@babel/traverse');
var ast = require('./ast.js');
require('prettier');
require('signale');
var format = require('./format.js');
require('path');
require('fs-extra');
require('art-template');
var fs$1 = require('./fs.js');
require('@babel/generator');
require('lodash/get');
var imports = require('./imports.js');
var exports$1 = require('./exports.js');
require('jsx-ast-utils');
require('strip-indent');
require('js-string-escape');
var jsx = require('./jsx.js');
require('to-vfile');
require('unified');
require('remark-parse');
require('remark-frontmatter');
require('remark-slug');
require('remark-parse-yaml');
require('unist-util-find');
require('unist-util-visit');
require('humanize-string');
require('lodash/flatten');
var mdast = require('./mdast.js');
var fp = require('lodash/fp');
var lodash = require('lodash');



exports.codeFromNode = ast.codeFromNode;
exports.valueFromTraverse = ast.valueFromTraverse;
exports.format = format.format;
exports.compiled = fs$1.compiled;
exports.touch = fs$1.touch;
exports.getFullImports = imports.getFullImports;
exports.getImportsVariables = imports.getImportsVariables;
exports.getExportsVariables = exports$1.getExportsVariables;
exports.componentName = jsx.componentName;
exports.propFromElement = jsx.propFromElement;
exports.removeTags = jsx.removeTags;
exports.sanitizeCode = jsx.sanitizeCode;
exports.getParsedData = mdast.getParsedData;
exports.headingsFromAst = mdast.headingsFromAst;
exports.parseMdx = mdast.parseMdx;
Object.defineProperty(exports, 'mergeWith', {
	enumerable: true,
	get: function () {
		return fp.mergeWith;
	}
});
Object.defineProperty(exports, 'get', {
	enumerable: true,
	get: function () {
		return lodash.get;
	}
});
